sympy/tensor/array/tests/test_array_comprehension.py

```diff
@@ -49,3 +49,18 @@ def test_array_comprehension():
     raises(ValueError, lambda: b.tolist())
     raises(ValueError, lambda: b.tomatrix())
     raises(ValueError, lambda: c.tomatrix())
+
+def test_arraycomprehensionmap():
+    a = ArrayComprehensionMap(lambda i: i+1, (i, 1, 5))
+    assert a.doit().tolist() == [2, 3, 4, 5, 6]
+    assert a.shape == (5,)
+    assert a.is_shape_numeric
+    assert a.tolist() == [2, 3, 4, 5, 6]
+    assert len(a) == 5
+    assert isinstance(a.doit(), ImmutableDenseNDimArray)
+
+    # tests about lambda expression
+    assert ArrayComprehensionMap(lambda: 3, (i, 1, 5)).doit().tolist() == [3, 3, 3, 3, 3]
+    assert ArrayComprehensionMap(lambda i: i+1, (i, 1, 5)).doit().tolist() == [2, 3, 4, 5, 6]
+    raises(ValueError, lambda: ArrayComprehensionMap(lambda i, j: i+j, (i, 1, 5)).doit())
+    raises(ValueError, lambda: ArrayComprehensionMap(lambda i: i+1, (i, 1, k)))
```

Comment:
This last one should not raise an exception. It looks OK. Replace it with:
```python
expr = ArrayComprehensionMap(lambda i: i+1, (i, 1, k))
assert expr.doit() == expr
```
This means that it cannot be expanded, because `k` is symbolic, therefore `.doit()` has no effect.

Reply:
If you overwrite `.func` like written above, `.subs( ... )` should work again.